crates/pipeline_manager/src/db/mod.rs

```diff
@@ -717,8 +989,154 @@ impl Storage for ProjectDB {
+                WHERE p.id = $1
+                    AND p.program_id = prog.id
+                    AND ac.pipeline_id = p.id
+                    AND ac.connector_id = c.id
+            ) as x",
+            &[&pipeline_id.0, &prev_revision],
+        )
+        .await?;
+        let nothing_changed = change.get::<_, i64>(0) == 0;
+        let committed_before = prev_revision > 0;
+        if committed_before && nothing_changed {
+            return Err(DBError::RevisionNotChanged.into());
+        }
+
+        // Copy all pipeline data to history tables
+        //
+        // TODO(performance): In theory the following inserts could all run in
+        // parallel with async but I couldn't figure out how to make it work
+        // with the args :/
+        txn.execute(
+            "INSERT INTO program_history SELECT $1 as revision, * FROM program p WHERE id = $2",
+            &[&revision, &program.program_id.0],
+        )
+        .await?;
+        txn.execute(
+            "INSERT INTO pipeline_history SELECT $1 as revision, * FROM pipeline p WHERE id = $2",
+            &[&revision, &pipeline_id.0],
+        )
+        .await?;
+        txn.execute(
+            "INSERT INTO connector_history SELECT $1 as revision, c.* FROM connector c, attached_connector ac WHERE ac.pipeline_id = $2 AND ac.connector_id = c.id",
+            &[&revision, &pipeline_id.0],
+        )
+        .await?;
+        txn.execute(
+            "INSERT INTO attached_connector_history SELECT $1 as revision, * FROM attached_connector ac WHERE pipeline_id = $2",
+            &[&revision, &pipeline_id.0],
+        ).await?;
+
+        // Update the revision of the pipeline object
+        txn.execute(
+            "UPDATE pipeline SET last_revision = $1 WHERE id = $2 AND tenant_id = $3",
+            &[&revision, &pipeline_id.0, &tenant_id.0],
+        )
+        .await?;
+
+        txn.commit().await?;
+
+        Ok(Revision(revision))
+    }
+
+    async fn get_last_committed_pipeline_revision(
+        &self,
+        tenant_id: TenantId,
+        pipeline_id: PipelineId,
+    ) -> AnyResult<PipelineRevision> {
+        let row: Option<Row> = self
+            .pool
+            .get()
+            .await?
+            .query_opt(
+                "SELECT last_revision FROM pipeline WHERE id = $1 AND tenant_id = $2",
+                &[&pipeline_id.0, &tenant_id.0],
+            )
+            .await?;
+
+        if let Some(row) = row {
+            let revision = Revision(
+                row.get::<_, Option<i64>>(0)
+                    .ok_or(DBError::NoRevisionAvailable(pipeline_id))?,
+            );
+            let pipeline = self
+                .get_committed_pipeline_by_id(tenant_id, pipeline_id, revision)
+                .await?;
+            // unwrap() is ok here -- we don't allow to version anything without a program:
+            let program_id = pipeline.program_id.unwrap();
```

Comment:
Should we use .expect() with the listed comment instead of unwrap? So that if a panic happens unintentionally we get the explanation in the logs too?